lightning/src/util/sweep.rs

```diff
@@ -463,39 +508,92 @@ where
 		self.sweeper_state.lock().unwrap().best_block
 	}
 
-	fn regenerate_spend_if_necessary(
-		&self, sweeper_state: &mut SweeperState,
-	) -> Option<Transaction> {
-		let cur_height = sweeper_state.best_block.height;
-		let cur_hash = sweeper_state.best_block.block_hash;
-		let filter_fn = |o: &TrackedSpendableOutput| {
-			if o.status.is_confirmed() {
-				// Don't rebroadcast confirmed txs.
-				return false;
-			}
+	/// Regenerates and broadcasts the spending transaction for any outputs that are pending
+	pub async fn regenerate_and_broadcast_spend_if_necessary_async(&self) -> Result<(), ()> {
+		// Collect spendable output descriptors.
+		let respend_descriptors_clones: Vec<SpendableOutputDescriptor>;
+		let respend_descriptors: Vec<&SpendableOutputDescriptor>;
+		{
+			let mut sweeper_state = self.sweeper_state.lock().unwrap();
 
-			if o.status.is_delayed(cur_height) {
-				// Don't generate and broadcast if still delayed
-				return false;
+			// Prevent concurrent sweeping.
+			if sweeper_state.sweep_pending {
+				return Ok(());
 			}
 
-			if o.status.latest_broadcast_height() >= Some(cur_height) {
-				// Only broadcast once per block height.
-				return false;
+			let cur_height = sweeper_state.best_block.height;
+			let filter_fn = |o: &TrackedSpendableOutput| {
+				if o.status.is_confirmed() {
+					// Don't rebroadcast confirmed txs.
+					return false;
+				}
+
+				if o.status.is_delayed(cur_height) {
+					// Don't generate and broadcast if still delayed
+					return false;
+				}
+
+				if o.status.latest_broadcast_height() >= Some(cur_height) {
+					// Only broadcast once per block height.
+					return false;
+				}
+
+				true
+			};
+
+			// Clone first, otherwise we can't take references that outlive the lock.
+			respend_descriptors_clones =
```

Comment:
Discussed offline - copy not necessary. Can re-select the descriptors in the second locked block.